### ISSUE TITLE
Is `autocast` needed with FSDP2?

### ISSUE THREAD

**Issue by garrett361** [pytorch/torchtitan#700]
Hi, is it necessary to wrap the forward pass in `autocast` when using FSDP2? I noticed that the `torchtitan` training loop does not.

If I wrap in `torch.autocast(device_type="cuda", dtype=torch.bfloat16)` my matmuls will be `bfloat16`, but my softmaxes (say) will be in `float32`. This behavior requires the autocast wrapper:
```python
t = torch.randn(100, device="cuda", dtype=torch.bfloat16)

with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
    out = t.softmax(dim=-1)

out.dtype # torch.float32

# Without autocast:
t.softmax(dim=-1).dtype # torch.bfloat16
```  
This is the usual way to do DDP or non-distributed mixed-precision training.

It seems to me that this behavior is lost in the `torchtitan` training loop which doesn't use the `autocast` [context manager](https://github.com/garrett361/torchtitan/blob/3247841423429faf37bdf6918204350db293e482/train.py#L308-L314). Is this not true? Does FSDP2 somehow still perform the upcast for the usual upcasted amp ops like softmax?  Not seeing how it might do so, and can't test easily at the moment.

I believe I correctly understand that `MixedPrecisionPolicy` controls the `dtype`s that weights are held in, reductions are performed in, and whether to cast a given module's outputs to a certain `dtype`, but that is all orthogonal to the dispatcher flags that `autocast` controls, IIUC.

Relates to #600 and #591. Also, I believe [OLMo uses autocast with FSDP](https://github.com/allenai/OLMo/blob/9c677c90cc881c37787c71373836d6889ad4de4a/olmo/train.py#L799-L809), but that is FSDP1 last time I checked.

CC @awgu

**Comment by awgu** [comment]
The behavior with FSDP2's native mixed precision is the same as FSDP1's. You need to use autocast manager if you want that softmax-in-fp32 behavior.

FSDP's native mixed precision is simply:
1. `param_dtype`: what dtype to all-gather parameters in?
2. `cast_forward_inputs`: should FSDP cast the module forward inputs to `param_dtype`?
3. `reduce_dtype`: what dtype to reduce-scatter gradients in?

The only casts that FSDP can provide are from 1, 2, and 3. There is no magic at the forward/backward operator level.
